soc: riscv: esp32c3: initial support for esp32c3 soc

uLipe opened the pull request:
Adds initial support for esp32c3 soc including its devkitm  board.

What is working:

* Basic kernel functionalities;
* Console log via printk;
* Generic samples such hello world and philosophers.

uLipe committed:
soc: esp32c3: added initial soc support files for esp32c3

by adding the soc specific files such: soc initialization code,
linker scripts and support for esp32c3 devkitm

Signed-off-by: Felipe Neves <ryukokki.felipe@gmail.com>
Signed-off-by: Felipe Neves <felipe.neves@espressif.com>

galak commented on dts/riscv/espressif/esp32c3.dtsi:
```diff
@@ -0,0 +1,32 @@
+		ranges;
+		
+		sram0: memory@3fc7c000 {
+			compatible = "espressif,esp32c3";
```
This seems like a odd compatible.  Any reason this isn't `mmio-sram`?

uLipe replied:
Probably this was inherited when I imported some lines from esp32 dts, I corrected it. 

Thanks!

katsuster commented on arch/riscv/Kconfig:
```diff
@@ -90,6 +90,12 @@ config RISCV_SOC_INTERRUPT_INIT
+	help
+	  Enable SOC-based pointer register initialization
+	  (call __soc_get_gp_initial_value when initializing a thread)
+
```
I think this is not SoC support. It's better to separate another PR or commit.

uLipe replied:
Thanks @katsuster I split this section in a separate commit.

katsuster commented on arch/riscv/core/thread.c:
```diff
@@ -47,6 +47,10 @@ void arch_new_thread(struct k_thread *thread, k_thread_stack_t *stack,
+#ifdef CONFIG_RISCV_SOC_INIT_GP_VALUE
+	stack_init->gp = __soc_get_gp_initial_value();
+#endif
+
```
Same as "config RISCV_SOC_INIT_GP_VALUE".

uLipe replied:
This one, as well

katsuster commented on soc/riscv/esp32c3/soc.c:
```diff
@@ -0,0 +1,189 @@
+
+	/* Start Zephyr */
+	z_cstart();
+
```
I guess this code want to avoid re-init C stack or bss areas in standard boot sequence (_PrepC() function).
However _PrepC does not only initialize the stack but also PMP, SoC depended initialization.

My concern is that can this code execute correctly even if changing implementation of _PrepC() in the future?

katsuster replied:
Unfortunately, Xtensa (ESP32) and RISC-V is not same structure.
Xtensa does not have _PrepC() so no one can add codes. But RISC-V has _PrepC() and someone expect (Ex. PMP) to run initialization at there. If exp32c3 can guarantee to initialize all feature/modules at this _start function, it's no problem. But I think it is very uncomfortable for future esp32c3 maintainer...

What is current _PrepC() problem? If that is only BSS clear, shall we add new CONFIG for avoiding BSS clear?

uLipe replied:
This make total sense, added the _PrepC at end of SoC specific startup code, delaying a bit the zeroing of the bss does not cause problem on the C3 special init code.

Thanks for the suggestion @katsuster

uLipe committed:
arch: riscv: added support for custom initialization of gp register

Plus added implementation for esp32c3 SoC.

Signed-off-by: Felipe Neves <ryukokki.felipe@gmail.com>
Signed-off-by: Felipe Neves <felipe.neves@espressif.com>

uLipe committed:
dts: riscv: espressif: correct sram0 compatible label to mmio-sram

And removes the odd esp32c3 competible from it

Signed-off-by: Felipe Neves <ryukokki.felipe@gmail.com>
Signed-off-by: Felipe Neves <felipe.neves@espressif.com>

uLipe committed:
drivers: timer: esp32c3: add esp32c3 systimer driver to CODEOWNERS

Also added maintainer to the entry

Signed-off-by: Felipe Neves <ryukokki.felipe@gmail.com>
Signed-off-by: Felipe Neves <felipe.neves@espressif.com>

katsuster commented on soc/riscv/esp32c3/soc.c:
```diff
@@ -0,0 +1,193 @@
+		/* Zero out BSS.  Clobber _bss_start to avoid memset() elision. */
+		(void)memset(&_bss_start, 0, (&_bss_end - &_bss_start) * sizeof(_bss_start));
+		__asm__ __volatile__("" : : "g"(&_bss_start) : "memory");
+
```
z_bss_zero() is not suitable?

uLipe replied:
Sure, this was also inherited from esp32 startup code, I added the z_bss_zero in place.

uLipe committed:
soc: riscv: esp32c3: add z_bss_zero in startup code

replaces the plain memcpy to zero the bss with
the standard zephyr function that does this task.

Signed-off-by: Felipe Neves <ryukokki.felipe@gmail.com>
Signed-off-by: Felipe Neves <felipe.neves@espressif.com>

uLipe committed:
soc: riscv: esp32c3: adds _PrepC to the startup code

The _PrepC() function is the standard risc-v way
of zephyr entry point, so let it call the z_cstart instead
of calling this function directly.

Signed-off-by: Felipe Neves <ryukokki.felipe@gmail.com>
Signed-off-by: Felipe Neves <felipe.neves@espressif.com>

uLipe committed:
soc: riscv: esp32c3: use the new esp_rom prefix

For esp32c3 related ROM located functions instead
of esp32c3_rom.

Signed-off-by: Felipe Neves <ryukokki.felipe@gmail.com>
Signed-off-by: Felipe Neves <felipe.neves@espressif.com>

uLipe committed:
west: update hal_espressif revision

In west manifest to support esp32c3 when using
west install / update

Signed-off-by: Felipe Neves <ryukokki.felipe@gmail.com>
Signed-off-by: Felipe Neves <felipe.neves@espressif.com>

uLipe committed:
esp32c3_devkitm: doc: add espressif toolchain

Instructions to build enviroment setup

Signed-off-by: Felipe Neves <ryukokki.felipe@gmail.com>
Signed-off-by: Felipe Neves <felipe.neves@espressif.com>

Commit:
west: update hal espressif

This new revision brings fix to the CI when
building espressif chips examples.

Signed-off-by: Felipe Neves <felipe.neves@espressif.com>

cwshu commented:
@uLipe
Based on my #36235 patch, I think we don't need to init GP register at thread init so that
the commit 'arch: riscv: ... gp register' is not needed. 
Could you also help to try #36235 patch on esp32c3? Any suggestion about this patch is welcome.

uLipe commented:
@cwshu , If your patch gets merged in the upstream before this one, I surely can remove this commit, otherwise I prefer to keep the commit until this happens because without the gp properly set the esp32c3 will not work correctly since it uses the gp in a specific reserved area.

cwshu commented:
@uLipe Thanks for your help. I could know that my patch about gp register also works on esp32c3 soc.

uLipe commented:
Rebased to main branch, now this one could be merged because it will not break esp32 anymore.

cc: @sylvioalves  @galak  @nashif
